KEP-3280: Guarantee PodDisruptionBudget When Preemption Happens (#3755)
kubernetes/enhancements: keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

This design proposal suggests adding a field `PreemptionPolicy` in the `PodDisruptionBudget` (PDB) API to indicate 
whether or not make `PodDisruptionBudget` guaranteed during the scheduler preemption process, in order to meet the 
needs of users who do not want PDB violations to occur during scheduler preemption due to some reasons such as 

> high availability of services(https://github.com/kubernetes/kubernetes/issues/91492#issuecomment-1029484252).

Author: Thanks. Added.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

PodDisruptionBudget (PDB) is used to limit the number of concurrent disruptions that your application experiences, 
allowing for high availability. Users can set the field '.spec.maxUnavailable' or '.spec.minAvailable' to declare 
the current minimum availability or maximum unavailability to be maintained after eviction. 

However, there is currently an issue where the kube-scheduler does not strictly guarantee PDBs during the preemption 

> process. The scheduler supports PDBs when preempting pods, but the adherence to PDBs is best effort. The scheduler 

Reviewer: Does the scheduler do this by inspecting PDBs and replicating the PDB controller logic, or by preempting pods via the /evictions endpoint?

Reviewer 2: replicating logic

Author: I thinks scheduler not just replicates the PDB controller logic, it inspects the PDB status and preempt the pods via the /pod/delete api.

https://github.com/kubernetes/kubernetes/blob/724497cda467b72bf56271637ec5edb00c19cedf/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L289-L335

Do you suggest to remove it via /evictions endpoint? @liggitt

Reviewer 2: We should not use the /evictions endpoint, as that will again respect PDB, which we need to violate at least for system priority.

Maybe there is chance to refactor some of the PDB logic so that it can be reused by scheduler. We should aim to do that.

Reviewer: > We should not use the /evictions endpoint, as that will again respect PDB, which we need to violate at least for system priority.

scheduler should use the evictions endpoint to preempt a pod while strictly honoring PDB

Reviewer 2: Ah, right. So we will have two deletion paths.

We need to make sure the scheduler cache works correctly if /evictions rejects the deletion.

Author: Yes. When the user declares the need for strict adherence to PDB, in order to avoid concurrent issues maybe arising from the controller, the final decision will still need to be made by the /evictions endpoint. However, if the user does not make such a declaration, pods will continue to be directly removed through the use of the remove method, as per the current implementation.

Reviewer 3: I think it might be useful to call api eviction also in the best effort case and if that fails fallback to the deletion.

It seems that the main problem of https://github.com/kubernetes/kubernetes/issues/91492 is the slow disruption controller reaction to the `status.disruptionsAllowed` which can race with the scheduler and cause incorrect disruptions.

So in the best effort case:
- if `disruptionAllowed > 0` the eviction will work and the api will decrement the PDB status, and the scheduler should watch and pick up this change almost instantly
- if `disruptionAllowed == 0` then we will fallback to deletion and the race should not occur since we are already at minimum

Also, if we start using the eviction API in all cases, we can sort the pods by ready status, and prefer the unready pods. The eviction would acknowledge a recently added field [unhealthyPodEvictionPolicy](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy) in PDBs which is relevant when `disruptionAllowed == 0`, and do least ammount of damage.

Author: > if disruptionAllowed == 0 then we will fallback to deletion and the race should not occur since we are already at minimum.

Do you mean when the value of disruptionAllowed is 0, the /eviction endpoint will return a response that rejects this option, and the scheduler will need to fallback to deletion as an alternative. Right?

> The eviction would acknowledge a recently added field [unhealthyPodEvictionPolicy](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy) in PDBs which is relevant when disruptionAllowed == 0, and do least ammount of damage.

Thanks, we have been working to minimize the impact of preemption on healthy running services, and unhealthyPodEvictionPolicy are a good option, though they may need to be considered for integration and use after they are released in beta.

Reviewer 3: > Do you mean when the value of disruptionAllowed is 0, the /eviction endpoint will return a response that rejects this option, and the scheduler will need to fallback to deletion as an alternative. Right?

yes

Reviewer 3: @denkensk can you please mention this under the BestEffort policy?

Author: Done @atiratree

Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1
I deployed a service in a cluster and to ensure its high availability, 
I created PDB for this deployment and set '.spec.minAvailable' to 3. I want the PDB to be 
guaranteed even in case of scheduler preemption.

#### Story 2
I created a Tensorflow distributed job that requires a minimum of 5 workers running.
I created PDB for the job and set '.spec.minAvailable' to 5. I want the PDB to be 
guaranteed even in case of scheduler preemption to ensure the stability of the whole job.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->
If a user sets a PodDisruptionBudget (PDB) for some low-priority pods and sets the `PreemptionPolicy` 
to `RequiredNotPreempted`, high-priority pods will not be able to violate the PDB and preempt 

> these pods during the scheduling process. This may result in high-priority pods being unable 

Author: Done.

to schedule while low-priority pods continue to run normally, which may make troubleshooting 
more complex for the user. To address this, additional logging or event descriptions will be 
added in the implementation to clearly inform the user of the reason why preemption did not occur.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> A new field named `PreemptionPolicy` will be added to `PodDisruptionBudgetSpec`.

Reviewer: Are PDB authors the right people to make this decision, or is the PriorityClass author the right person? It seems wrong to let a namespace-constrained user who is only allowed to create low priority pods to consume cluster resources and block high priority pods from preempting.

At first glance, I would have expected preemption behavior to be specified in PriorityClass, as a "pods with this priority class should only be preempted when allowed by PDB" option

Author: > It seems wrong to let a namespace-constrained user who is only allowed to create low priority pods to consume cluster resources and block high priority pods from preempting.

Thanks. Your comments is greatly appreciated. We had also considered adding a field in the PriorityClass to indicate a behavior where preemption isn't allowed with violating PDB in the draft of our previous design document. https://docs.google.com/document/d/1g2wj8S04M0yTy2t4l4GzrdARtxBpkBTfkJHw0oqBvXM/edit#

However, as you have mentioned, the potential conflict between the creators of PDB and PriorityClass, who may have different priorities(cluster scope and namespace scope), has been overlooked and may result in high priority pods being blocked from preemption in the cluster. 

So, based on the above issues and your suggestions, we can address this by adding a field in the PriorityClass, which would allow for this behavior to be defined and managed consistently by the cluster administrator in the PriorityClass, rather than relying on the creators of PDB and PriorityClass potentially having conflicting priorities. 

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy

	//  PreemptedPolicy describes a policy for determining if the pod can be preempted
        //  by other higher priority pod.
	PreemptedPolicy *core.PreemptedPolicy
}

//  PreemptedPolicy describes a policy for determining if the pod can be preempted
//  by other higher priority pod.
type PreemptedPolicy string

const (
	// PreemptedAllowedByPDB indicates that a pod will not be preempted by higher
	// priority pods if it would result in a violation of the PDB.
	PreemptedAllowedByPDB PreemptedPolicy = "PreemptedAllowedByPDB"
	// PreemptedAlways indicates that a pod will be preempted by higher priority
	// pods at any time.
	PreemptedAlways PreemptedPolicy = "PreemptedAlways"
)
```
WDYT? @liggitt @alculquicondor

Reviewer 2: That looks better, but I don't like the name.

Maybe PreemptionTargetPolicy?

Author: Thanks. In terms of naming, I do not have a clear preference as long as it can accurately expresses the policy for selecting a pod as a victim when it is preempted.

Author: @liggitt  Could you please help to take a look if it looks good to you? If so, I will modify the design document according to this idea? Thanks.

Reviewer: Placing the field on priorityClass makes sense, though the name is still confusing to me. Would something like `DisruptionPolicy` be more clear?

```go
// DisruptionPolicy specifies how pods with this priority class may be disrupted when being preempted.
DisruptionPolicy *DisruptionPolicy
```

for possible values, it seems like the current behavior is not quite "Always", it *tries* to honor PDB in a best-effort / if-possible sort of way. Would names that reflect that be clearer?
```
// - PodDisruptionBudgetBestEffort
// - PodDisruptionBudgetStrict
```

Reviewer 3: +1 for moving the responsibilities to the priorityClass. Does not seems right for application owners affect priorities that they do not own.

It might be preferable to mention the preemption in the field name, as there might be other reason for disruption than preemption that we might want to configure per priority class in the future. Or maybe just name it `PodDisruptionBudgetPolicy` to be completely clear?

violation of the PDB will be allowed during the scheduler preemption process. The 
default value for this field will be `PreferNotPreempted`, which is consistent with 
current behavior. Users will be able to set this field to determine whether or not 
PDB violation will be allowed during scheduler preemption. Additionally, in order 
to avoid system stability issues, even if the victim's PDB is not allowed to be 
violated during the preemption process, the preempter with priority ClassName of 
`system-cluster-critical` or `system-node-critical` will still be able to violate 
the victim's PDB during preemption.

```go
// PodDisruptionBudgetSpec is a description of a PodDisruptionBudget.
type PodDisruptionBudgetSpec struct {
	MinAvailable *intstr.IntOrString
	Selector *metav1.LabelSelector
	MaxUnavailable *intstr.IntOrString
    
  // PreemptionPolicy indicates whether pdb can be violated 
  // during preemption. Pods with priority ClassName of
  // system-cluster-critical or system-node-critical 
  // can violate the victim's pdb during preemption.
  // Defaults to PreferNotPreempted if unset.
  // +optional

>   PreemptionPolicy *PreemptionPolicy

Reviewer 2: Maybe we can be a bit more explicit:

```yaml
permitPreemptionBy: SystemCritical|HigherPriority
```

cc @liggitt as SIG Auth TL and API reviewer

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- These cases will be added in the existed integration tests:
  - Feature gate enable/disable tests
  - During scheduling, `PreemptionPolicy` in `PodDisruptionBudget` works as expected
  - Verify no significant performance degradation

- `k8s.io/kubernetes/kubernetes/test/integration/scheduler/preemption/preemption_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=TestPreemption
- `k8s.io/kubernetes/test/integration/scheduler_perf/scheduler_perf_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=BenchmarkPerfScheduling

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- These cases will be added in the existed e2e tests:

>   - Feature gate enable/disable tests

Reviewer 2: We don't generally have this in E2E tests, as it's not easy to change configurations of the cluster.

But also we might be ok with just integration tests, as there are no more dependencies that scheduler and apiserver

Author: Ok. Updated.

      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->
- 2023-01-19: Initial KEP

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> 1. Add `PreemptLowerPriorityWithoutViolatePDB` as an option in the `PreemptionPolicy`. 

Reviewer 2: So this alternative is about giving the policy control to the preemptor, rather than the victim pod. Is that correct? If so, it's not very clear in the description.

Author: Yes. it's the policy for the preemptor. Additional explanations have been added to the design document.

   When the preemptor is set to `PreemptLowerPriorityWithoutViolatePDB`, the pods that would violate 
   PDBs will be excluded when selecting victims. However, if we want to guarantee some pods' PDBs, 
   we need to modify the `PreemptionPolicy` for all other pods to `PreemptLowerPriorityWithoutViolatePDB`. 
   The cost of this operation may be relatively large, and it's a little strange. Also, the more natural 
   place to describe PDB behavior is in the PDB spec.

> 2. Add `NotAllowViolatePDB` as an option in the `PreemptionPolicy`. When the pods are set to `NotAllowViolatePDB`, 

Reviewer 2: Same question as the one above.

They actually sound like the same alternative, but with slightly different names.

Author: NO. The option1 is the policy for the preemptor, option2 is the policy for the victim. Additional explanations have been added to the design document.

   they are not preemptible in violation of PDBs. We need to exclude these pods when preemption occurs. However, 
   to ensure the security and stability of the cluster, pods whose priority class is `system-cluster-critical` 
   or `system-node-critical` can violate the victim's PDB. But it's not consistent with the positioning of the 
   `PreemptionPolicy`, it affects the victims not the preempter. Also, the more natural place to describe PDB 

>    behavior is in the PDB spec.

Reviewer: I think I disagree with this... this makes one part of the system contradict priorityclass... pods and things within a namespace should not be able to take action to make themselves take resources away from higher priority pods contrary to the cluster administrator's configuration

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

In order to address the issue mentioned above, a new field `DisruptionPolicy` will 
be added to `PriorityClass`. Users will be able to set this field to 
determine whether or not PDB violation will be allowed during scheduler preemption. The 
default value for this field will be `PodDisruptionBudgetBestEffort`, which is 
consistent with current behavior.  Additionally, in order to avoid system stability 
issues, even if the victim's PDB is not allowed to be violated during the preemption 

> process, the preempter with priority ClassName of `system-cluster-critical` or `system-node-critical` 

Reviewer 3: I wonder if the system classes shouldn't be enforced by the API? Maybe by having a `SystemPriority` preemption policy that would be used only for these and would override the `DisruptionPolicy`?

I know, it seems like we are running in a circle about who has the most rights. So it is question if we want to have this hard coded or customizable?

Reviewer 2: But that's pretty much the semantics of the field `PreemptionPolicy`.

I think the point of this KEP is that giving all the control to the higher priority class is not enough for some users.

Reviewer 2: I think we need to reconsider having `PreemptLowerPriorityWithoutViolatePDB` in the higher priority pod.

And if a pod is not important enough to block preemption, it probably shouldn't have a PDB.

Author: >  I think we need to reconsider having PreemptLowerPriorityWithoutViolatePDB in the higher priority pod.
> And if a pod is not important enough to block preemption, it probably shouldn't have a PDB.

It's true that what you said makes sense, but this may require the user to configure strictly in this way, which would result in a high cost for the user. For example, he needs to check the PDB of all pods with Priority Class. In this way, it might be better to configure only the Priority Class that he needs to protect, with a lower cost.

Reviewer 2: Please clarify the semantics of your proposal when the user sets both PreemptionPolicy and DisruptionPolicy, in other words, the precedence.

Reviewer 4: PriorityClass used to cover the semantics of "preemptor" only, and now it starts to take "preemptee" into consideration. So +1 to define the precedence.

In a real-world case, disruption is more unacceptable than unscheduled. So I'd suggest always prioritizing DisruptionPolicy over PreemptionPolicy. This can also guide the users to opt-in (hard) disruption policy for their workloads to avoid disruption.

> I wonder if the system classes shouldn't be enforced by the API? Maybe by having a SystemPriority

I'd like to make the semantics explicitly defined in API as well, but to ease understanding the semantics, IMO we should apply on the "preemptee" side:

- PDBStrict (means "once scheduled, nobody can preempt me")
- PDBStrictExceptForSystemPods (means "once scheduled, only high-priority pods that are system pods can preempt me")

> I think we need to reconsider having `PreemptLowerPriorityWithoutViolatePDB` in the higher priority pod.

I'd hold off adding more policies to the "preemptor" side, as it will make the semantics of disruption/preemption ambiguous. If workloads don't want to be preempted, they should declare DisruptionPolicy explicitly; instead of expecting the preemptors to have "mercy" on them.

Reviewer 2: what you are calling "mercy" could be a compromise to the stability of the system. Higher priority pods might have higher priority because they are critical to the system. How do we get the power back to those pods?

Reviewer 4: > How do we get the power back to those pods?

(Not quite sure I understand the power correctly) I mean with the introduction of disruption policy, the semantics is already not that intuitive, so I'd suggest keeping the **preemptor** side simple, unless it's a must-need. Otherwise, we may end up with a super completed semantics of preemption. Suppose we introduce other hard disruption policies, like `ABCStrict`, `XYZStrict`, and we want to follow the same pattern of "PreemptLowerPriorityWithoutViolatePDB", we will end up with introducing more preemption policies like `PreemptLowerPriorityWithoutViolateABC`, `PreemptLowerPriorityWithoutViolateXYZ` and `PreemptLowerPriorityWithoutViolateABCAndXYZ` - this is way too complicated.

So if we set a tone on day 1 to accommodate preemption customization on "preemptee" side _only_, it may make the semantics more maintainable, extensible, and easily understood.

Reviewer 4: > But we already have policies from the preemptor side.

yes, but those are pretty fundamental ones (PreemptLower and PreemptNever).

> So we keep all configurations in the preemptor side, we are already minimizing confusion.

But we're adding semantics (disruption policy) on preemptee side. I doubt if we can afford more complex policy combinations on **_both_** sides.

Author: > Please clarify the semantics of your proposal when the user sets both PreemptionPolicy and DisruptionPolicy, in other words, the precedence.

Added in the design: The PreemptionPolicy is used to describe the behavior of the pods associated with the PriorityClass during preemption as the preemptor. The DisruptionPolicy is used to describe the policy of the pods associated with the PriorityClass during preemption as the victim. During a preemption process, if the preemptor is configured with the PreemptionPolicy and the victim is configured with the DisruptionPolicy, the DisruptionPolicy takes priority over the PreemptionPolicy.

Author: > But we already have policies from the preemptor side. So we keep all configurations in the preemptor side, we are already minimizing confusion.

I still feel that this approach would result in a high cost. Currently, we cannot guarantee that all users will use PDB effectively. That is, if a pod is not important enough to block preemption, it probably shouldn't have a PDB. 

In this sense, for users who need to use our feature, if they want to protect a certain service, they need to set all other PriorityClasses to Strict, whereas only setting their own service's corresponding PriorityClass to Strict would be easier.

Reviewer 2: It is easier, yes, but that's not what I'm debating here. My points are:
- It adds complexity when used in combination with PreemptionPolicy (thanks for clarifying the precedence, though)
- It gives too much power to low priority classes. How can a higher priority class take the power back? Consider a case where a service is more important than a Pod with PDB and DisruptionPolicy, but not important enough to be system-critical.

Reviewer 4: > It gives too much power to low priority classes

Yes it is, so I do appreciate the discussion earlier that it's agreed this disruption policy should be imposed to PrirotyClass (cluster-scoped) instead of PDB (namespaced). Cluster SRE/Admin should be responsible for PriorityClass's management, and when they opt-in this feature, they should know the implication of "XYZStrict".

> How can a higher priority class take the power back?

We define clearly define it in disruption policy. For example:

```yaml
disruptionPolicy:
  XYZStrictExcluding: # means no PC can disrupt me except for the followings
  - system-cluster-critical
  - system-node-critical
  - a-particular-high-priority
```

Reviewer 2: Maybe it should be a numeric value. Every pod with priority higher than the given value is allowed to preempt the Workload. And that value can't be higher than the value for system-node-critical.

It's a bit more resilient to changes to the set of PriorityClasses.

Reviewer 4: > Maybe it should be a numeric value.

Makes sense. The spec can be like:

```yaml
disruptionPolicy:
- name: XYZStrict
  excludePriorityValueGT: <numeric value>
```

or, more expressible

```yaml
disruptionPolicy:
- name: XYZStrict
  priorityClassExclusionSelector:
  ...
```

Reviewer 2: ```yaml
allowPreemptionByPriorityGreaterThan: <value>
```

When it's not set, it's equal to the priority of the class  +1, which gives the current behavior.

Author: ```yaml
disruptionPolicy:
- name: PDBStrict
   allowPreemptionByPriorityGreaterThan: <numeric value>
```
+1 it's more resilient

Reviewer 2: I don't like how the proposal adds "strict" and then the next field is an exception. And in combination with preemptionPolicy it becomes a nightmare to explain which field takes precedence.

Reviewer 2: And why does it need to be an array?

Author: > I don't like how the proposal adds "strict" and then the next field is an exception. 

How about this one? @alculquicondor  It will be easier and clearer.

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy
        
        // If not set, it equals to the `Value + 1`
	AllowDisruptionByPriorityGreaterThan int32
}
```

> And why does it need to be an array?

Consider if there will be other policies, but there shouldn't be.

Reviewer 4: > in combination with preemptionPolicy it becomes a nightmare to explain which field takes precedence.
> And why does it need to be an array?

It's the same question indeed. If there is no need to have multiple disruption policies, I agree flattening them is more clear:

```
disruptionPolicy: PDBStrict
allowPreemptionByPriorityGreaterThan: <value>
```

Author: How about just use `AllowDisruptionByPriorityGreaterThan` directly ?  
```yaml
allowDisruptionByPriorityGreaterThan: <value>
```

Greater than Value, less than or equal to AllowDisruptionByPriorityGreaterThan is Strict. 
Greater than AllowDisruptionByPriorityGreaterThan is BestEffort.

Reviewer 2: Maybe `GreaterThanOrEqual` would be better so that you can use node-system-critical easily instead of users having to calculate node-system-critical-1.

Reviewer 4: > How about just use AllowDisruptionByPriorityGreaterThan directly?

Do you mean with a single field and hardcode underlying scop to PDB? or two fields?

Author: Updated by the suggestion from https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094884915 
Pls take a look. @alculquicondor @Huang-Wei

Author: also /cc @liggitt

Reviewer 5: I wouldn't try to design the ideal API at the KEP time, we can always update the KEP based on the api review.

Reviewer 4: +1 to leave API spec to API review, but I do that having a struct `DisruptionPolicy` (and under that define a "PDB" _explicitly_) will be more extensible.

will still be able to violate the victim's PDB during preemption.

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy

	// DisruptionPolicy specifies how pods with this priority class 
	// may be disrupted when being preempted by a higher priority pod.
	// Defaults to PodDisruptionBudgetBestEffort if unset.
	// +optional

> 	DisruptionPolicy *core.DisruptionPolicy

Reviewer 2: Also explain what happens when this field is set in combination to `PreemptionPolicy` in the other `PriorityClass`. Which one takes precedence?

Reviewer 2: Would `PreemptionDisruptionPolicy` be more accurate? This is specifically about the preemption behavior, not other kinds of disruptions.

@liggitt for other suggestions

}
```

Currently, two simple policies are provided in this version. One is `PodDisruptionBudgetBestEffort` 
which indicates that the scheduler will try to avoid violating the PDB during preemption, 
but it cannot be guaranteed. The other is `PodDisruptionBudgetStrict` which indicates that the 
PDB will not be violated during scheduler preemption, as noted above, if the preempter 
has a priority ClassName of `system-cluster-critical` or `system-node-critical`, it may 
still potentially violate the victim's PDB. New policies may be added in the future based 
on the need, but it will not be discussed in this article at this time.

```go
// DisruptionPolicy specifies how pods with this priority class 
// may be disrupted when being preempted.
type DisruptionPolicy string

const (
	// PodDisruptionBudgetStrict indicates when being preempted by a higher priority pod, 
	// strict adherence to the PodDisruptionBudget of the pod with this priority class is required.
	// But it the preempter with these priority classes of `system-cluster-critical` or `system-node-critical` 
	// will still be able to violate this pod's PDB during preemption.

> 	PodDisruptionBudgetStrict PreemptedPolicy = "PodDisruptionBudgetStrict"

Reviewer 2: `SystemCriticalOnly`

Reviewer 3: to consider: https://github.com/kubernetes/enhancements/pull/3755/files#r1091298934

Reviewer 4: nit: `PDBStrict DisruptionPolicy = ...`

	
	// PodDisruptionBudgetBestEffort indicates when being preempted by a higher priority pod, 
	// an effort is made to adhere to the PodDisruptionBudget of the pod with this priority class
	// as much as possible.

> 	PodDisruptionBudgetBestEffort PreemptedPolicy = "PodDisruptionBudgetBestEffort"

Reviewer 2: just `BestEffort`

scheduler preemption process.

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: low-priority
value: 100
globalDefault: false
description: "This priority class should be used for XYZ service pods only."
disruptionPolicy: PodDisruptionBudgetStrict
```

The scheduler plugin `defaultpreemption` needs to check the policy set in the "preemptionPolicy" field 
when selecting victims.
- if it's set to `PodDisruptionBudgetBestEffort`, the implementation will remain consistent with the existing behavior,
  meaning that the scheduler will try to select victims whose PDBs are not violated by preemption, but 
  if no such victims are found, preemption will still happen and lower priority pods will be removed 
  despite their PDBs being violated.
- if it's set to `PodDisruptionBudgetStrict`, there are two options:
  - If the preempter's priority ClassName is `system-cluster-critical` or `system-node-critical` the implementation 
    will be the same as `PodDisruptionBudgetBestEffort`.
  - If not, the scheduler will check if the victim' PDBs will be violated when selecting victims
    - if violate the victims' PDBs, this victim will not be selected as candidates.

>     - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint.

Reviewer 2: Please expand on what happens if the scheduler sends a call through `/evictions` but the pod is not deleted. Would the incoming pod be able to choose to preempt a different pod in the next iteration?

What would cause the scheduler to retry this pod if `/evictions` rejects the deletion? Since there is no pod being deleted, is there any event the scheduler can react to?

Reviewer 3: one of the ways would be react to the HTTP code of the eviction request. 429 (`TooManyRequests`) comes to mind but it also gets returned when `DisruptionsAllowed == 0` so we would either have to distinguish that via the message, or locally somehow, or distinguish & update the error responses in the eviction API

Author: If it responds `200 OK`,  it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses  `429 Too Many Requests` or `500 Internal Server Error`, the scheduler will output an error log and terminate the preemption process (the same as the current behavior of deletion failure).

Author: > Would the incoming pod be able to choose to preempt a different pod in the next iteration?

Maybe don't need，In addition to the apiserver service network issue, failure may only occur in the event of concurrent issues between the scheduler and controller.

Reviewer 2: > and terminate the preemption process

My question is about what happens after that. The pod should be entering the unschedulable queue. What would bring it back into the active queue? Or would it stay in the active/backoff queue?

Reviewer 3: FYI, I meant in these two cases it is possible to retry eviction later.

- https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L301
- https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L390

not sure if it is necessary to mention 500 error code in the enhancement as there are other possible error codes as well

Author: > My question is about what happens after that. The pod should be entering the unschedulable queue. What would bring it back into the active queue? Or would it stay in the active/backoff queue?

get it. 

You are right, if `/evictions` interface is called during the preemption process, but the responses is `429 Too Many Requests`. Assuming no additional events arrive, the preemptor cannot be moved to the activeQ until it is flushed. 

So I think we should not terminate the preemption process directly, but choose another victim among the candidate victims(remove the previous victims from victimsMap and run pickOneNodeForPreemption again) to preempt until it succeeds or there are no more candidates.

      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->
- 2023-01-19: Initial KEP
- 2023-01-28: Move the responsibilities from `PodDisruptionBudget` to `PriorityClass`

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> 1. Add `PreemptLowerPriorityWithoutViolatePDB` as an option in the `PreemptionPolicy` of the preempter. 

Reviewer 2: I think we need to reconsider this option more thoroughly. Most clusters would have a small set of PriorityClasses that would make this option feasible.

Although, one downside of this option is that every low priority class can restrict preemption of higher priority classes that set `PreemptLowerPriorityWithoutViolatePDB`

Reviewer 4: > one downside of this option is that every low priority class can restrict preemption of higher priority classes that set PreemptLowerPriorityWithoutViolatePDB

That's why I think workloads should declare their disruption policy explicitly, if they really want to be not interrupted. Otherwise, scheduler would assume they're adhering to the numeric pc-value based preemption.

  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
<!-- /toc -->

## Release Signoff Checklist

<!--
**ACTION REQUIRED:** In order to merge code into a release, there must be an
issue in [kubernetes/enhancements] referencing this KEP and targeting a release
milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)              
of the targeted release**.

For enhancements that make changes to code or processes/procedures in core
Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
Signoff checklist to be completed.

Check these off as they are completed for the Release Team to track. These
checklist items _must_ be updated for the enhancement to be released.
-->

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)

> - [ ] (R) KEP approvers have approved the KEP status as `implementable`

Reviewer 2: You need to mark these boxes to target the release.

Author: Marked.

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

This design proposal suggests adding a field `AllowDisruptionByPriorityGreaterThanOrEqual` in the `PriorityClass` API 

> to explicitly indicates  that `PodDisruptionBudget` of the pods corresponding to this priorty class can only be 

Reviewer 5: ```suggestion
to explicitly indicate that `PodDisruptionBudget` of the pods corresponding to this priorty class can only be 
```

violated by pods with the priority value greater than or equal to the value of `AllowDisruptionByPriorityGreaterThanOrEqual` 

> during the scheduler preemption process, in order to meet the needs of users who do not want PDB violations to occur 

Reviewer 2: nit
```suggestion
during the scheduler preemption process.  This proposal allows cluster administrators to define PriorityClasses that restrict PDB violations during preemption to satisfy the needs for high availability of services
```

allowing for high availability. Users can set the field `.spec.maxUnavailable` or `.spec.minAvailable` to declare 
the current minimum availability or maximum unavailability to be maintained after eviction. 

However, there is currently an issue where the kube-scheduler does not strictly guarantee PDBs during the preemption 
process. The scheduler supports PDBs when preempting pods, but the adherence to PDBs is best effort. The scheduler 
attempts to select victims whose PDBs are not violated during preemption, but if no such victims are found, preemption 
will still take place, resulting in the removal of lower-priority pods despite their PDBs being violated.

PodDisruptionBudgets (PDBs) are frequently used for stability and the possibility of violating a PDB during the preemption 
process is not acceptable for certain users. As such, it is beneficial to provide users with the option to choose if they 
want the PDB to be guaranteed during preemption or not.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> - Provide an option to configure whether the scheduler needs to make `PodDisruptionBudget` 

Reviewer 2: Maybe make explicit that this option is available to cluster administrators only.

And then clarify that it's a non goal to let application developers directly influence preemption behavior.

Author: Updated.

> - Provide an optional option to configure whether the scheduler needs to make `PodDisruptionBudget` 

Reviewer 5: ```suggestion
- Provide an option to configure whether the scheduler needs to make `PodDisruptionBudget` 
```

  guaranteed when preemption happens.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> I deployed a service in a cluster and to ensure its high availability, 

Reviewer 5: Stories are usually written in 3rd person, ie. User deploys a service...

Author: Done

For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

If a user sets a PodDisruptionBudget (PDB) for some low-priority pods and sets the `AllowDisruptionByPriorityGreaterThanOrEqual` 
to `PriorityClass`, high-priority (less than the value of allowDisruptionByPriorityGreaterThanOrEqual) pods will not be able 
to violate the PDB and preempt these pods during the scheduling process. This may result in high-priority (less than the value of allowDisruptionByPriorityGreaterThanOrEqual) pods being unable to schedule while low-priority pods continue to run normally. 

Although the above situations may arise, due to the fact that PriorityClass is created and managed by 
cluster administrators with no permission for application owners to perform actions, administrators 
are able to uniformly configure according to the requirements. Additionally, implementation will 
include the addition of additional logging or event descriptions to clearly inform the user of the 

> reason why preemption did not occur.

Reviewer 2: Also system-critical will always be able to preempt.

Author: Added.

Reviewer 6: Good to explicitly document, but that should not require special coding, since that is already the highest priority ?

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

In order to address the issue mentioned above, a new field `AllowDisruptionByPriorityGreaterThanOrEqual` will 
be added to `PriorityClass`. Users will be able to set this field to indicate that the PodDisruptionBudget of 
the pods associated with this priority class can only be violated during scheduler preemption by other pods 
with a priority value greater than or equal to AllowDisruptionByPriorityGreaterThanOrEqual.
The default value for this field will be `Value` + 1, which is consistent with current behavior.  

```go
type PriorityClass struct {
  metav1.TypeMeta
  metav1.ObjectMeta
  Value int32
  GlobalDefault bool
  Description string
  PreemptionPolicy *core.PreemptionPolicy

  // AllowDisruptionByPriorityGreaterThanOrEqual indicates that the PodDisruptionBudget of the pods 
  // associated with this priority class can only be violated during scheduler preemption by other pods 

>   // with a priority value greater than or equal to AllowDisruptionByPriorityGreaterThanOrEqual.

Reviewer 2: The value must be lower than the one for node-system-critical

Author: Updated.

>   // Defaults to Value+1 if unset.

Reviewer 2: We shouldn't default to it. Instead, there should be an interpretation for null.

This is because users might want to change the `Value` field without having to worry about `AllowDisruptionBy...` if they didn't initially change it.

Author: Yes, you are right, I overlooked the possibility that the user might modify the priority value.
Added the interpretation for null.

  // +optional

>   AllowDisruptionByPriorityGreaterThanOrEqual int32

Reviewer 2: ```suggestion
  AllowDisruptionByPriorityGreaterThanOrEqual *int32
```

}
```

The `AllowDisruptionByPriorityGreaterThanOrEqual` field in PodSpec will be populated during pod admission, 
similarly to how the PriorityClass Value is populated. Storing the `AllowDisruptionByPriorityGreaterThanOrEqual`
field in the pod spec has several benefits:

1. The scheduler does not need to be aware of PiorityClasses, as all relevant information is in the pod.
2. Mutating PriorityClass objects does not impact existing pods.

```go
// PodSpec is a description of a pod.
type PodSpec struct {
  PriorityClassName string
  Priority *int32
  PreemptionPolicy *PreemptionPolicy

+ // AllowDisruptionByPriorityGreaterThanOrEqual indicates that the PodDisruptionBudget of the pods 
+ // associated with this priority class can only be violated during scheduler preemption by other pods 
+ // with a priority value greater than or equal to AllowDisruptionByPriorityGreaterThanOrEqual. When 
+ // Priority Admission Controller is enabled, it prevents users from setting this field. 
+ // The admission controller populates this field from PriorityClassName.
+ // Defaults to Value+1 if unset.
+ // +optiona
+ AllowDisruptionByPriorityGreaterThanOrEqual *int32

> )

Reviewer 2: ```suggestion
}
```

```

The scheduler plugin `defaultpreemption` needs to check the value set in the `AllowDisruptionByPriorityGreaterThanOrEqual` field 
when selecting victims.
- if the priority of the preemptor is greater than or equal to the value of `AllowDisruptionByPriorityGreaterThanOrEqual` in victim pod, 
  the implementation will remain consistent with the existing behavior, meaning that the scheduler will try to select victims whose PDBs 
  are not violated by preemption, but if no such victims are found, preemption will still happen and lower priority pods will be removed 
  despite their PDBs being violated.
- if the priority of the preemptor is less than the value of `AllowDisruptionByPriorityGreaterThanOrEqual` in victim pod, 
  the scheduler will check if the victim' PDBs will be violated when selecting victims
  - if violate the victims' PDBs, this victim will not be selected as candidates.
  - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint. If it responds `200 OK`, 
    it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses

>     `429 Too Many Requests`, the scheduler will output an error log and choose another victim among the candidate victims to preempt 

Reviewer 2: this sounds expensive to recalculate within the scheduling cycle, as we might even have to choose victims on a different node.

Should we just abort this cycle and retry after putting the pod in backoff?

Author: > This sounds expensive to recalculate within the scheduling cycle, as we might even have to choose victims on a different node.

We might just have to recalculate this one node and don't need to recalculate totally again. We can reuse the `nodesToVictims`  from the `findCandidates` phase and the intermediate results from the `SelectCandidate` phase. 

> Should we just abort this cycle and retry after putting the pod in backoff?

It's OK. But assuming no other relevant events arrive, the preemptor cannot move to activeQ until it has been flushed after 30s. It's maybe to long.

Reviewer 2: Sounds good. We might have to adjust some behavior once you have a prototype, but we can leave that for code reviews.

Reviewer 3: I think it is a bit confusing in the KEP that we mention only code 429, when other error codes can happen as well, but I suppose we can leave that to the code reviews as well.

Reviewer 2: Other errors should be handled like we handle any other error in scheduler: abort the scheduling cycle. No changes there.

Reviewer 3: It might be useful to select another victim when for example pdbs are misconfigured or pods are migrating [eviction.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L215). But the error code is too generic, and I am not sure if we want to rely that much on the API for such behaviours.

In the original behaviour branch, the fallback to delete should happen at all times, unless we get 200 from eviction. To be backwards compatible and ensure best effort case.

with a priority value greater than or equal to AllowDisruptionByPriorityGreaterThanOrEqual. At the same time, to 
prevent situations where core components or necessary add-ons cannot be scheduled due to the inability to violate 
PDBs, the value of `AllowDisruptionByPriorityGreaterThanOrEqual` cannot be greater than the priority value of 
`system-cluster-critical` and `system-node-critical.priority`.

```go
type PriorityClass struct {
  metav1.TypeMeta
  metav1.ObjectMeta
  Value int32
  GlobalDefault bool
  Description string
  PreemptionPolicy *core.PreemptionPolicy

  // AllowDisruptionByPriorityGreaterThanOrEqual indicates that a PodDisruptionBudget set for pods associated 
  // with this priority class can only be violated by pods with a priority value greater than or equal to 
  // AllowDisruptionByPriorityGreaterThanOrEqual during a preemption process. The value of AllowDisruptionByPriorityGreaterThanOrEqual 
  // cannot be greater than the priority value of system-cluster-critical or system-node-critical.
  // A null value indicates that the PodDisruptionBudget is allow to be disrupted by any other pods with a higher priority.
  // +optional

>   AllowDisruptionByPriorityGreaterThanOrEqual *int32

Author: FYI https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094875982

preemption as the victim. During a preemption process, if the preemptor is configured with the `PreemptionPolicy` and the victim is 
configured with the `AllowDisruptionByPriorityGreaterThanOrEqual`, the `AllowDisruptionByPriorityGreaterThanOrEqual` takes priority over 
the `PreemptionPolicy`.

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

> ##### Unit tests

Reviewer 7: Didn't read the codes in priorityClass, don't we need a bit of tests there? Or I just missed in the KEP.

Author: The test codes relevant PriorityClass is in these pkgs: `pkg/scheduler/apis/config/v1` and `k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption`. I will add the additional tests in these pkgs.

motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

`PodDisruptionBudget` (PDB) is used to limit the number of concurrent disruptions that your application experiences, 
allowing for high availability. Users can set the field `.spec.maxUnavailable` or `.spec.minAvailable` to declare 
the current minimum availability or maximum unavailability to be maintained after eviction. 

However, there is currently an issue where the kube-scheduler does not strictly guarantee PDBs during the preemption 
process. The scheduler supports PDBs when preempting pods, but the adherence to PDBs is best effort. The scheduler 
attempts to select victims whose PDBs are not violated during preemption, but if no such victims are found, preemption 
will still take place, resulting in the removal of lower-priority pods despite their PDBs being violated.

PodDisruptionBudgets (PDBs) are frequently used for stability and the possibility of violating a PDB during the preemption 

> process is not acceptable for certain users. As such, it is beneficial to provide users with the option to choose if they 

Reviewer 6: I think its important to note that preemption is done by both kubelet and kube-scheduler. The kubelet pre-emption honors PDB but kube-scheduler doesn't and this KEP is about kube-scheduler pre-emption

and make progress.
-->
- Let the application developers influence preemption behavior directly.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1
User deployed a service in a cluster and to ensure its high availability, 
User created PDB for this deployment and set `.spec.minAvailable` to 3. User wanted the PDB to be 
guaranteed even in case of scheduler preemption.

#### Story 2
User created a Tensorflow distributed job that requires a minimum of 5 workers running.
User created PDB for the job and set `.spec.minAvailable` to 5. User wanted the PDB to be 

> guaranteed even in case of scheduler preemption to ensure the stability of the whole job.

Reviewer 6: is the main difference between the two stories, about tensorflow ?

PDBs, the value of `AllowDisruptionByPriorityGreaterThanOrEqual` cannot be greater than the priority value of 
`system-cluster-critical` and `system-node-critical.priority`.

```go
type PriorityClass struct {
  metav1.TypeMeta
  metav1.ObjectMeta
  Value int32
  GlobalDefault bool
  Description string
  PreemptionPolicy *core.PreemptionPolicy

  // AllowDisruptionByPriorityGreaterThanOrEqual indicates that a PodDisruptionBudget set for pods associated 
  // with this priority class can only be violated by pods with a priority value greater than or equal to 
  // AllowDisruptionByPriorityGreaterThanOrEqual during a preemption process. The value of AllowDisruptionByPriorityGreaterThanOrEqual 
  // cannot be greater than the priority value of system-cluster-critical or system-node-critical.
  // A null value indicates that the PodDisruptionBudget is allow to be disrupted by any other pods with a higher priority.
  // +optional
  AllowDisruptionByPriorityGreaterThanOrEqual *int32
}
```

The `AllowDisruptionByPriorityGreaterThanOrEqual` field in PodSpec will be populated during pod admission, 

> similarly to how the PriorityClass Value is populated. Storing the `AllowDisruptionByPriorityGreaterThanOrEqual`

Reviewer 6: i agree this seems unnecessary and duplicative detail to push into pod. This belongs in the  Priortyclass only. I hope we can revisit this

field in the pod spec has several benefits:

> 1. The scheduler does not need to be aware of PiorityClasses, as all relevant information is in the pod.

Reviewer 6: why is this important ?

  - if violate the victims' PDBs, this victim will not be selected as candidates.
  - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint. If it responds `200 OK`, 
    it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses
    `429 Too Many Requests`, the scheduler will output an error log and choose another victim among the candidate victims to preempt 
    until it succeeds or there are no more candidates.

### PreemptionPolicy vs AllowDisruptionByPriorityGreaterThanOrEqual

The `PreemptionPolicy` is used to describe the behavior of the pods associated with the PriorityClass during preemption as the preemptor. 
The `AllowDisruptionByPriorityGreaterThanOrEqual` is used to describe the policy of the pods associated with the PriorityClass during 
preemption as the victim. During a preemption process, if the preemptor is configured with the `PreemptionPolicy` and the victim is 
configured with the `AllowDisruptionByPriorityGreaterThanOrEqual`, the `AllowDisruptionByPriorityGreaterThanOrEqual` takes priority over 

> the `PreemptionPolicy`.

Reviewer 6: We may rephrase this as PreemptionPolicy only applies wgeb priority is lower or higher

should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

This design proposal suggests adding a field `DisruptionPolicy` in the `PriorityClass` API to indicate 
whether or not make `PodDisruptionBudget` guaranteed during the scheduler preemption process, in order to meet the 
needs of users who do not want PDB violations to occur during scheduler preemption due to some reasons such as 
high availability of services(https://github.com/kubernetes/kubernetes/issues/91492#issuecomment-1029484252). 
At the same time, to prevent situations where core components or necessary add-ons cannot be scheduled due to the 
inability to violate PDBs, priority ClassName of `system-cluster-critical` or `system-node-critical` will have the 

> exemption and not be required to comply with this configuration, and the behavior will be consistent with the current one.

Reviewer 8: Hmm - I'm somewhat on the fence with this feature. Without this feature, if I was running higher-prioirty pods I was guaranteed that my pods will get scheduled, because they can preempt something else.

Now, having a higher-priority pods doesn't buy me much, because if the cluster is full (and there is no place to schedule anything else) and lower-level PDB targets are disrupted to their levels, I'm stuck.

I'm not entirely sure I'm buying the usecase for this feature - if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

I think this change would introduce enormous confusion to users - I'm higher-priority, but I'm still not able to run in the cluster even though lower-priority stuff is running there.

This feature effectively means that at least `MinAvailable` replicas from each set of pods protected by PDB in those priority class effectively isn't in this priority class, but rather is in the highest-possible priority class (except from system-cluster-critical and system-node-critical).

That is opening the system for gaming the system:
- I'm running with the lowest priority-class, so let me mark this bit and set PDB target high enough and I will be guaranteed to run a huge number of my pods.

This makes this feature really risky imho. I would really like us to spend more time on figuring out other options we can address the main usecase for it.

Reviewer 4: > if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

Somehow this represents a semantics of "scheduled in a low-priroity, but once running it's a high-priority that can only bare minimum disruption". Stateful workloads that have associated (local) storage are of this type.

> I'm higher-priority, but I'm still not able to run in the cluster even though lower-priority stuff is running there.

Yes, it sounds the case. But in practice, only very restricted PriorityClasses are created in this way, under cluster SRE's control. In other words, priority value is not the only factor to impact whether a pod can preempt others.

> That is opening the system for gaming the system:

So when opt-in this feature, SRE admin should be cautious when creating this kind of PriorityClass, they should know what they're doing.

--

Providing a data point about gaming: I'd say the current priority/preemption system _can_ be gamed already:

- a high-priority pod H requests 2 CPUs, without PDB, scheduled on node A
- a low-priority pod L requests 2 CPUs, with PDB protected, scheduled on node B
- node A & B are both fully utilized
- now it comes in a system-critical pod S requesting 2CPUs. In this case, although pod H holds a higher priority than L, given scheduler honors PDB in a best-efforts manner, H gets preempted instead of L

Believe it or not, this is one reason some end-users don't use PDB in their prod env. So a side-product of this KEP is to give an option for SRE admin to disable this behavior to avoid abuse, although it will be still enabled by default for backwards compatibility.

Reviewer 2: Not sure I understand the problem. H should have a PDB if can't tolerate disruption. But if it's disrupted, the deployment would create another pod that should conquer over L.

> So a side-product of this KEP is to give an option for SRE admin to disable this behavior to avoid abuse.

How? Are you proposing a mechanism to do preemption without looking at PDBs?

Reviewer 4: > How? Are you proposing a mechanism to do preemption without looking at PDBs?

Yes, behave like one type of disruption policy. But I don't want to derail the topic here, where we should focus on the general idea of disruption policy, as well as different controls on PDB.

> the deployment would create another pod that should conquer over L.

There can be medium priorityclass in between, also the combination of PDB violated/not-violated on a node can be pretty complex. Take a step back, even if we _could_ prove eventually L will fail on this abuse game, it's still nonsense to me if it's based on the cost of so many extra steps (casdaded premeption, re-create pods).

Author: > Hmm - I'm somewhat on the fence with this feature. Without this feature, if I was running higher priority pods I was guaranteed that my pods will get scheduled, because they can preempt something else.

Maybe not. The value in priority class is not the only criteria that determines whether a Pod can be preempted. There is also another field called preemptionPolicy in PriorityClass that sets the preemption strategy. A PriorityClass with a high Value may not be able to preempt other lower priority Pods. https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class

> I'm not entirely sure I'm buying the use case for this feature - if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

It's a common occurrence in real production environments that some services or jobs have a lower priority and can be preempted, but once they reach the `MinAvailable`, every pod‘s importance increases and they can no longer be preempted arbitrarily. In addition to Stateful workloads as mentioned by Wei, there are also distributed training jobs for AI and BigData, which can run and be preempted by other high priority pods, but if they fall below MinAvailable, the entire task will fail directly, leading to a bigger impact.

> That is opening the system for gaming the system: I'm running with the lowest priority-class, so let me mark this bit and set PDB target high enough and I will be guaranteed to run a huge number of my pods.

Yes, this is why we placed the field in PriorityClass instead of PDB based on @liggitt 's suggestion. https://github.com/kubernetes/enhancements/pull/3755#discussion_r1083853660 . Only cluster administrators can create and configure PriorityClass, and they can decide which PriorityClass needs to be protected, avoiding any gaming behavior among regular users.

In addition, the objective of this design document is to clarify the behavior of PDBs during scheduling. In the past, users were unable to determine if their services would be preempted by the scheduler under PDBs. To ensure the integrity of service PDBs, simply increasing the value of PriorityClass is not effective and will only lead to low-priority tasks becoming high priority, thereby impacting the scheduling of pods with high priority.

Reviewer 8: > Maybe not. The value in priority class is not the only criteria that determines whether a Pod can be preempted. There is also another field called preemptionPolicy in PriorityClass that sets the preemption strategy. A PriorityClass with a high Value may not be able to preempt other lower priority Pods. https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class

I don't agree with this one. This is saying that "my pods can't preempt someone else" - I own the decision that my pods can't even preempt lower-priority pods. This is the right permission to have, as I don't affect higher-priority pods (I'm not saying you can't preempt me even if you're more important).

> Yes, this is why we placed the field in PriorityClass instead of PDB based on @liggitt 's suggestion. https://github.com/kubernetes/enhancements/pull/3755#discussion_r1083853660 . Only cluster administrators can create and configure PriorityClass, and they can decide which PriorityClass needs to be protected, avoiding any gaming behavior among regular users.

I spent some time thinking about it and I'm less sceptical about it now.
I would actually move what you have under "Notes/Constraints/Caveats (Optional)" section to "Risks and Mitigations" to explicitly describe the problem (low-prio pods becoming high-prio pods) and an explicit mitigation we have for that (giving this power only to administrator).

be added to `PriorityClass`. Users will be able to set this field to 
determine whether or not PDB violation will be allowed during scheduler preemption. The 
default value for this field will be `PodDisruptionBudgetBestEffort`, which is 
consistent with current behavior.  Additionally, in order to avoid system stability 
issues, even if the victim's PDB is not allowed to be violated during the preemption 
process, the preempter with priority ClassName of `system-cluster-critical` or `system-node-critical` 
will still be able to violate the victim's PDB during preemption.

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy

	// DisruptionPolicy specifies how pods with this priority class 
	// may be disrupted when being preempted by a higher priority pod.
	// Defaults to PodDisruptionBudgetBestEffort if unset.
	// +optional
	DisruptionPolicy *core.DisruptionPolicy
}
```

Currently, two simple policies are provided in this version. One is `PodDisruptionBudgetBestEffort` 
which indicates that the scheduler will try to avoid violating the PDB during preemption, 

> but it cannot be guaranteed. The other is `PodDisruptionBudgetStrict` which indicates that the 

Reviewer 8: Switch to `SystemCriticalOnly` as mentioned below.

Reviewer 8: And in all other places too (assuming that `SystemCriticalOnly` and `BestEffort` are names we're going with, which actually makes more sense to me, as it better reflects the reality).

Author: Sorry. I updated the design by the suggestion from https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094884915 
Pls take a look again.

```

The scheduler plugin `defaultpreemption` needs to check the policy set in the "preemptionPolicy" field 
when selecting victims.
- if it's set to `PodDisruptionBudgetBestEffort`, the implementation will remain consistent with the existing behavior,
  meaning that the scheduler will try to select victims whose PDBs are not violated by preemption, but 
  if no such victims are found, preemption will still happen and lower priority pods will be removed 
  despite their PDBs being violated.
- if it's set to `PodDisruptionBudgetStrict`, there are two options:
  - If the preempter's priority ClassName is `system-cluster-critical` or `system-node-critical` the implementation 
    will be the same as `PodDisruptionBudgetBestEffort`.
  - If not, the scheduler will check if the victim' PDBs will be violated when selecting victims
    - if violate the victims' PDBs, this victim will not be selected as candidates.
    - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint. If it responds `200 OK`, 
      it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses

>       `429 Too Many Requests` or `500 Internal Server Error`, the scheduler will output an error log and terminate the preemption process 

Reviewer 8: Shouldn't we retry here on other preemption candidates here?
@alculquicondor

Author: Yes. we can retry on other candidates. Updated by the https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094868040

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- These cases will be added in the existed integration tests:
  - Feature gate enable/disable tests
  - During scheduling, `DisruptionPolicy` in `PriorityClass` works as expected
  - Verify no significant performance degradation

- `k8s.io/kubernetes/kubernetes/test/integration/scheduler/preemption/preemption_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=TestPreemption
- `k8s.io/kubernetes/test/integration/scheduler_perf/scheduler_perf_test.go`: https://storage.googleapis.com/k8s-triage/index.html?test=BenchmarkPerfScheduling

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> N/A

Reviewer 8: Hmm - shouldn't we have an e2e test for this too?
[We already have tests for preemption and this new mode seems like a useful thing to validate too]

Author: Updated. We have an exist e2e test for preemption.

the feature gate, however kube-scheduler will not take the field into account.

###### What happens if we reenable the feature if it was previously rolled back?
1. The newly created PriorityClass will contain the field `DisruptionPolicy`.
2. The scheduler will check the value in the field `DisruptionPolicy` in `PriorityClass` 
   if preemption occurs during scheduling

###### Are there any tests for feature enablement/disablement?

<!--
The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

> No, unit and integration tests will be added.

Reviewer 8: FWIW - the only enablement/disablement test that I expect here is an api test described in the comment above.

[Do you have any other enablement/disablement tests on your mind (note that we're talking about tests that change the value of feature-gate in the middle of the test here) that would be useful here?]
